#!/usr/bin/env python

import os
import json
import sys
import traceback
import click
import numpy as np
from jtnn.reconstruct_eval import reconstruct as run_experiment

TORCH = True
try:
    import torch
except ImportError:
    TORCH = False


@click.command()
@click.option('--config-file', '-p', help="Path to the config file (json) that contains the parameters for the experiment.")
@click.option('--output-path', '-o', help="Location for saving the training results (model artifacts and output files).")
def main(config_file, output_path):
    print('Starting the training.')
    os.makedirs(output_path, exist_ok=True)
    try:
        # Read in any hyperparameters that the user passed with the training job
        # Depending on how you set the hyperparameters
        train_params = {}
        with open(config_file, 'r') as tc:
           train_params = json.load(tc)
        # the function below does all the data loading, run, validate and test the algo
        print("Launching with {}".format(train_params))
        run_experiment()


    except Exception as e:
        # Write out an error file. This will be returned as the failureReason in the
        # DescribeTrainingJob result.
        trc = traceback.format_exc()
        with open(os.path.join(output_path, 'failure'), 'w') as s:
            s.write('Exception during training: ' + str(e) + '\n' + trc)
        # Printing this causes the exception to be in the training job logs, as well.
        print('Exception during training: ' +
              str(e) + '\n' + trc, file=sys.stderr)
        # A non-zero exit code causes the training job to be marked as Failed.
        sys.exit(255)


if __name__ == '__main__':
    np.random.seed(42)
    if TORCH:
        torch.manual_seed(42)
    main()
    sys.exit(0)
